homeassistant/components/rituals_perfume_genie/switch.py

```diff
@@ -1,104 +1,77 @@
-    diffusers = await account.get_devices()
-
+    diffusers = hass.data[DOMAIN][config_entry.entry_id][DEVICES]
+    coordinators = hass.data[DOMAIN][config_entry.entry_id][COORDINATORS]
     entities = []
-    for diffuser in diffusers:
-        entities.append(DiffuserSwitch(diffuser))
+    for hublot, diffuser in diffusers.items():
+        coordinator = coordinators[hublot]
+        entities.append(DiffuserSwitch(diffuser, coordinator))
 
-    async_add_entities(entities, True)
+    async_add_entities(entities)
 
 
-class DiffuserSwitch(SwitchEntity):
+class DiffuserSwitch(SwitchEntity, DiffuserEntity):
     """Representation of a diffuser switch."""
 
-    def __init__(self, diffuser):
-        """Initialize the switch."""
-        self._diffuser = diffuser
-        self._available = True
-
-    @property
-    def device_info(self):
-        """Return information about the device."""
-        return {
-            "name": self._diffuser.data["hub"]["attributes"]["roomnamec"],
-            "identifiers": {(DOMAIN, self._diffuser.data["hub"]["hublot"])},
-            "manufacturer": MANUFACTURER,
-            "model": MODEL,
-            "sw_version": self._diffuser.data["hub"]["sensors"]["versionc"],
-        }
-
-    @property
-    def unique_id(self):
-        """Return the unique ID of the device."""
-        return self._diffuser.data["hub"]["hublot"]
+    def __init__(self, diffuser, coordinator):
+        """Initialize the diffuser switch."""
+        super().__init__(diffuser, coordinator, "")
+        self._is_on = self.coordinator.data[HUB][ATTRIBUTES][FAN] == ON_STATE
 
     @property
     def available(self):
         """Return if the device is available."""
-        return self._available
-
-    @property
-    def name(self):
-        """Return the name of the device."""
-        return self._diffuser.data["hub"]["attributes"]["roomnamec"]
+        return self.coordinator.data[HUB][STATUS] == AVAILABLE_STATE
 
     @property
     def icon(self):
         """Return the icon of the device."""
-        return ICON
+        return "mdi:fan"
 
     @property
     def extra_state_attributes(self):
         """Return the device state attributes."""
         attributes = {
-            "fan_speed": self._diffuser.data["hub"]["attributes"]["speedc"],
-            "room_size": self._diffuser.data["hub"]["attributes"]["roomc"],
+            "fan_speed": self.coordinator.data[HUB][ATTRIBUTES][SPEED],
+            "room_size": self.coordinator.data[HUB][ATTRIBUTES][ROOM],
         }
         return attributes
 
     @property
     def is_on(self):
         """If the device is currently on or off."""
-        return self._diffuser.data["hub"]["attributes"]["fanc"] == ON_STATE
+        return self._is_on
 
     async def async_turn_on(self, **kwargs):
         """Turn the device on."""
         await self._diffuser.turn_on()
+        self._is_on = True
+        self.schedule_update_ha_state()
```

Comment:
We can't use the sync api in async context. Use `self.async_write_ha_state` instead.